Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php

```diff
@@ -238,11 +265,92 @@ protected function resolveValue($node)
         if ($this->onlyMatchSiteNodes() && $node !== $siteNode) {
             return false;
         }
+        $resolvedNode = $this->nodeShortcutResolver->resolveShortcutTarget($node);
+        if ($resolvedNode === null) {
+            throw new NeosException(sprintf('Could not resolve shortcut target for node "%s"', $node->getPath()), 1414771137);
+        }
+
+        $node = $this->resolveShortcutNode($node);
+        if ($node instanceof UriInterface) {
+            return new ResolveResult('', UriConstraints::fromUri($node), null);
+        }
 
-        $routePath = $this->resolveRoutePathForNode($node);
-        $this->value = $routePath;
+        try {
+            $uriConstraints = $this->buildUriConstraintsForResolvedNode($node);
+        } catch (Exception\NoSiteException $exception) {
+            $this->systemLogger->debug('FrontendNodeRoutePartHandler resolveValue(): ' . $exception->getMessage());
+            return false;
+        }
+        return new ResolveResult('', $uriConstraints);
+    }
 
-        return true;
+    /**
+     * @param NodeInterface $node
+     * @return NodeInterface|Uri The original, unaltered $node if it's not a shortcut node. Otherwise the nodes shortcut target (a node or an URI for external & asset shortcuts)
+     * @throws NeosException
+     */
+    protected function resolveShortcutNode(NodeInterface $node)
+    {
+        $resolvedNode = $this->nodeShortcutResolver->resolveShortcutTarget($node);
+        if (is_string($resolvedNode)) {
+            return new Uri($resolvedNode);
+        }
+        if (!$resolvedNode instanceof NodeInterface) {
+            throw new NeosException(sprintf('Could not resolve shortcut target for node "%s"', $node->getPath()), 1414771137);
+        }
+        return $resolvedNode;
+    }
+
+    /**
+     * Builds UriConstraints for the given $node with:
+     * * a path constraint equal to the resolved URI path
+     * * domain specific constraints for nodes in a different Neos site
+     * * a path suffix corresponding to the configured "uriPathSuffix"
+     *
+     * @param NodeInterface $node
+     * @return UriConstraints
+     * @throws Exception\NoSiteException
+     */
+    protected function buildUriConstraintsForResolvedNode(NodeInterface $node): UriConstraints
+    {
+        $uriPath = $this->resolveRoutePathForNode($node);
+        $uriConstraints = UriConstraints::create()->withPath($uriPath);
+        $requestSite = $this->getCurrentSite();
+        if (!NodePaths::isSubPathOf(SiteService::SITES_ROOT_PATH, $node->getPath())) {
+            throw new Exception\NoSiteException(sprintf('The node at path "%s" is not located underneath the sites root path "%s"', $node->getPath(), SiteService::SITES_ROOT_PATH), 1604922914);
```

Comment:
Should this throw? because you could very well want to link to a node in another site? I think it should do that (like in the removed code from linking service)

Reply:
The exception is caught in `resolveValue()` and turned into a corresponding log entry (like we did it for many other parts in the handler). I added a doc comment to the `@throws` annotation